feat: add DaemonSet mount support for StorageClass

elijah-rou opened the pull request:
Implements DaemonSet deployment mode for JuiceFS mount pods when using StorageClass with mount sharing enabled. This provides better resource management and control over which nodes run mount pods.

Key features:
- Three deployment modes: per-pvc, shared-pod, and daemonset
- ConfigMap-based configuration for existing StorageClasses
- Node affinity support to control mount pod placement
- Reference tracking via DaemonSet annotations
- Automatic lifecycle management of DaemonSets
- Backward compatibility with existing deployments

Configuration:
- Enable via STORAGE_CLASS_DAEMONSET=true environment variable
- Configure node affinity via ConfigMap or StorageClass parameters
- MountSelector dynamically chooses the appropriate mount implementation

The implementation includes comprehensive unit tests and documentation to guide users through the configuration process.

elijah-rou committed:
feat: add DaemonSet mount support for StorageClass

Implements DaemonSet deployment mode for JuiceFS mount pods when using
StorageClass with mount sharing enabled. This provides better resource
management and control over which nodes run mount pods.

Key features:
- Three deployment modes: per-pvc, shared-pod, and daemonset
- ConfigMap-based configuration for existing StorageClasses
- Node affinity support to control mount pod placement
- Reference tracking via DaemonSet annotations
- Automatic lifecycle management of DaemonSets
- Backward compatibility with existing deployments

Configuration:
- Enable via STORAGE_CLASS_DAEMONSET=true environment variable
- Configure node affinity via ConfigMap or StorageClass parameters
- MountSelector dynamically chooses the appropriate mount implementation

The implementation includes comprehensive unit tests and documentation
to guide users through the configuration process.

elijah-rou committed:
docs: fix markdown linting issues

- Capitalize 'Mount Pod' and 'Mount Pods' consistently per enhanced-proper-names rule
- Fix all occurrences in daemonset-mount.md and mount-pod-configuration.md
- Comply with .markdownlint-cli2.jsonc configuration

jonoirwinrsa commented on pnpm-lock.yaml:
@elijah-rou i don't really understand - why does this file exist

elijah-rou replied:
md lint

jonoirwinrsa commented:
@elijah-rou is the idea to make a PR into Juice after we merge into our own?

elijah-rou commented:
> @elijah-rou is the idea to make a PR into Juice after we merge into our own?

yes

yaseenisolated commented on deploy/kubernetes/csi-daemonset-mount/daemonset-config.yaml:
```diff
@@ -0,0 +1,38 @@
+apiVersion: v1
+kind: ConfigMap
+metadata:
+  name: juicefs-daemonset-config
```
rename to juicefs-mountpod-config

yaseenisolated commented on deploy/kubernetes/csi-daemonset-mount/daemonset-config.yaml:
```diff
@@ -0,0 +1,38 @@
+            - "true"
+  
+  # Another StorageClass with different affinity
+  juicefs-sc-performance: |
```
this doesn't seem correct.

yaseenisolated commented on deploy/kubernetes/csi-daemonset-mount/daemonset.yaml:
```diff
@@ -0,0 +1,70 @@
+apiVersion: apps/v1
+metadata:
+  name: juicefs-csi-node
+  namespace: kube-system
```
needs to be in juicefs or whatever namespace we're configured to run in.

wesrobin commented on pkg/config/daemonset_config.go:
```diff
@@ -0,0 +1,128 @@
+	
+	// Skip if not using DaemonSet deployment
+	if !StorageClassShareMount || !StorageClassDaemonSet {
+		return nil
```
Shouldn't this be an error? If this is called from elsewhere and it doesn't error the assumption will be that the state is set

elijah-rou committed:
feat: add comprehensive mount pod configuration system with DaemonSet support

- Implement three mount modes: per-pvc, shared-pod, and daemonset
- Add ConfigMap-based configuration for per-StorageClass mount settings
- Support DaemonSet deployment with flexible nodeAffinity configuration
- Add automatic fallback from DaemonSet to shared pod when node scheduling fails
- Rename daemonset_config to mount_config for clarity
- Add node compatibility checking before DaemonSet scheduling
- Include comprehensive tests for mount selection and fallback behavior
- Update documentation with configuration examples and best practices

This allows users to optimize mount pod deployment based on their specific
needs without modifying existing StorageClasses, with automatic fallback
ensuring workloads never fail due to mount pod scheduling issues.

elijah-rou committed:
feat: add CI workflows for building and pushing JuiceFS CSI images to ECR

- Add ci_release.yml for production releases with auto-versioning
- Add ci_update_dev.yml for development and PR builds
- Configure Depot for multi-arch builds (amd64/arm64)
- Build CSI driver, mount images (CE/EE), and dashboard
- Push to ECR with region-specific naming (virginia/london)
- Use specific EE image version ee-5.2.8-1305e8c

elijah-rou committed:
refactor: simplify mount configuration by removing STORAGE_CLASS_DAEMONSET flag

- Remove separate STORAGE_CLASS_DAEMONSET environment variable
- DaemonSet mode is now automatically available when STORAGE_CLASS_SHARE_MOUNT is enabled
- Changed LoadDaemonSetNodeAffinity to assert mount sharing is enabled rather than conditionally check
- Updated documentation to reflect the simplified configuration
- Updated all tests to work without the removed flag

This simplifies the configuration model - when mount sharing is enabled,
users can choose between shared-pod (default) and daemonset modes via ConfigMap
without needing a separate environment variable.
